### PR TITLE
Add LazyBlob helper class to lazy load file from filesystem

### PR DESCRIPTION
**Pull request by Aschen**
## Description

This PR add a new helper class `LazyBlob` which aim to prevent loading huge file into RAM.

This new class can be used instead of a `Blob` with the `commit` function.

Fix https://github.com/huggingface/huggingface.js/issues/68

```js
const lazyBlob = await LazyBlob.create("./package.json");

await commit({
  repo,
  title: "Some commit",
  credentials: {
    accessToken: "<access token>",
  },
  operations: [
    {
      operation: "addOrUpdate",
      content: lazyBlob,
      path: "package.json",
    },
  ],
});
```

### CODE REVIEW

File: packages/hub/src/lib/LazyBlob.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
+import { createReadStream } from "fs";
+import { open } from "fs/promises";
+import type { FileHandle } from "fs/promises";
```
Comment (coyotte508) [inline packages/hub/src/lib/LazyBlob.ts:3]:
```suggestion
import { createReadStream } from "node:fs";
import { open } from "node:fs/promises";
import type { FileHandle } from "node:fs/promises";
```

File: packages/hub/src/lib/LazyBlob.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	}
+
+	async text(): Promise<string> {
+		const buffer = (await this.arrayBuffer()) as Buffer;
+
+		return buffer.toString("utf8");
+	}
+
+	stream(): ReadableStream {
+		const stream = createReadStream(this.path);
```
Comment (coyotte508) [inline packages/hub/src/lib/LazyBlob.ts:80]:
```suggestion
		return Readable.toWeb(createReadStream(this.path));
```

And:

```ts
import { Readable } from "node:stream";
import type { ReadableStream } from "node:stream/web";
```

**Review by coyotte508 — COMMENTED** [review]
This is going to be really cool when ready!

My main request is that from the outside, it should behave exactly as a `Blob`. No need for `.init()` or `dispose()`.

File: packages/hub/src/lib/LazyBlob.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
+import { createReadStream } from "fs";
+import { open } from "fs/promises";
+import type { FileHandle } from "fs/promises";
+
+export class LazyBlob {
```
[the remark this answers is not in the document]
Comment (Aschen) [reply packages/hub/src/lib/LazyBlob.ts:5]: I'm not sure about this since the `since` methods will have different signatures (promise and not promise)

**Comment by Aschen** [comment]
@coyotte508 I can get ride of `init` with the private constructor and static `create` method as you suggest but for the dispose the only option would be to `open` the file when reading and slice and then `close` it.

Here it's DX vs performances so I will let you choose :+1:

File: packages/hub/src/lib/LazyBlob.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+		await this.file.close();
+	}
+
+	async slice(start: number, end: number): Promise<Blob> {
+		if (this.file === null) {
+			throw new Error("LazyBlob has not been initialized");
+		}
+
+		const size = end - start;
```
[the remark this answers is not in the document]
Comment (Aschen) [reply packages/hub/src/lib/LazyBlob.ts:49]: About negative values for `end` and `start`, althought it's possible to do this, it's not trivial because it would require in some case to read the file 2 times. 

For example on a file of size 10:
 - `slice(5, -2)`: read from 5 to 10, then from 0 to 1
 
To be honest, I'm not sure it's necessary to mimic the whole `Blob` class since we could just stick to the usecase we have. If the feature is needed later then we can add it at the cost of more complexity .

File: packages/hub/src/lib/LazyBlob.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+		const slice = await this.file.read(Buffer.alloc(size), 0, size, start);
```
[the remark this answers is not in the document]
Comment (Aschen) [reply packages/hub/src/lib/LazyBlob.ts:50]: I have the same comment about minimal working `LazyBlob` for our usecase. If needed, this feature could be added later.

File: packages/hub/src/lib/LazyBlob.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (coyotte508) [reply packages/hub/src/lib/LazyBlob.ts:5]: With the other comments, `.size` and `.slice` should be able to be generated synchronously without promises:

- size is given at init
- For `.slice`, a `new LazyBlob(path, start, end)`;

The static function to create the blob can be: `const fileSize = await ...; return new LazyBlob(path, 0, fileSize)`

**Comment by coyotte508** [comment]
> @coyotte508 I can get ride of `init` with the private constructor and static `create` method as you suggest but for the dispose the only option would be to `open` the file when reading and slice and then `close` it.
> 
> Here it's DX vs performances so I will let you choose :+1: 

The most calls will be three I think, on large files:

- a `.slice(0, 512)` to get a sample to send to `/preupload`
- a `.stream()` to compute the `Sha`
- Then whatever `fetch` will call (probably `.stream()`)

On smaller files, it will be two calls: one `.slice(0, 512)` and one `.arrayBuffer()`.

Given that, I'd rather prioritize DX on this one :)

**Comment by Aschen** [comment]
> Given that, I'd rather prioritize DX on this one ! The overhead should be low. Maybe an optimization would be to directly load small files in memory, if we really care about perf.

I agree with that :+1:  Also, better have more syscall to open/close than unused file descriptor :see_no_evil:

File: packages/hub/src/lib/LazyBlob.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Aschen) [reply packages/hub/src/lib/LazyBlob.ts:6]: Let me implements that :+1:

File: packages/hub/src/utils/sha256.ts
Comment (coyotte508) [inline packages/hub/src/utils/sha256.ts:1]: This file should be able to be reverted

File: packages/hub/src/lib/commit.ts
Comment (coyotte508) [inline packages/hub/src/lib/commit.ts:1]: This file should now also be able to be reverted (maybe just remove the todo comment)

File: packages/hub/src/lib/LazyBlob.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+import { Readable } from "node:stream";
+
+export class LazyBlob extends Blob {
+	static async create(path: string, start?: number, end?: number): Promise<LazyBlob> {
+		const lazyBlob = new LazyBlob(path, start, end);
+
+		await lazyBlob.init();
+
+		return lazyBlob;
+	}
```
Comment (coyotte508) [inline packages/hub/src/lib/LazyBlob.ts:13]:
```suggestion
	static async create(path: string): Promise<LazyBlob> {
		const { size } = await stat(path);
		
		const lazyBlob = new LazyBlob(path, 0, size);

		return lazyBlob;
	}
```

That way `LazyBlob.start` and `LazyBlob.end` will always be defined. It'll remove the need for `LazyBlob.totalSize` too.

File: packages/hub/src/lib/LazyBlob.ts [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+		return this.totalSize;
+	}
+
+	get type(): string {
+		return "";
+	}
+
+	slice(start = 0, end = this.size): LazyBlob {
+		const slice = new LazyBlob(this.path, start, end);
```
Comment (coyotte508) [inline packages/hub/src/lib/LazyBlob.ts:47]:
```suggestion
		const slice = new LazyBlob(this.path, this.start + start, Math.min(this.start + end, this.end));
```

We should also throw `new TypeError("Unsupported negative start/end on LazyBlob.slice")`  if `start < 0` or `end < 0`, in case we don't want to support this behavior (we don't really need to support it for now)

File: packages/hub/src/lib/LazyBlob.ts [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+		return this.slice(0, this.size);
+	}
+
+	async arrayBuffer(): Promise<ArrayBuffer> {
+		const start = this.start || 0;
+		const end = this.end || this.size;
+		const size = Math.abs(end - start);
+		const cappedSize = size > this.size ? this.size : size;
+
+		const slice = await this.execute((file) => file.read(Buffer.alloc(cappedSize), 0, cappedSize, 0));
```
Comment (coyotte508) [inline packages/hub/src/lib/LazyBlob.ts:62]:
```suggestion
		const slice = await this.execute((file) => file.read(Buffer.alloc(this.size), 0, this.size, this.start));
```

With the other requested change on the blob creation, it should be possible to simplify the code to this

File: packages/hub/test.js
Comment (coyotte508) [inline packages/hub/test.js:1]: I think this belongs to another PR :)

**Review by coyotte508 — COMMENTED** [review]
Getting close!

File: packages/hub/test.js
Comment (Aschen) [reply packages/hub/test.js:1]: Oopsi :see_no_evil:

File: packages/hub/src/lib/LazyBlob.ts [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+	stream(): ReadableStream {
+		return Readable.toWeb(createReadStream(this.path));
+	}
+
+	private async init(): Promise<void> {
+		const { size } = await this.execute((file) => file.stat());
+
+		this.totalSize = size;
+	}
```
Comment (coyotte508) [inline packages/hub/src/lib/LazyBlob.ts:76]: This can be removed

File: packages/hub/src/lib/LazyBlob.ts [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+	async text(): Promise<string> {
+		const buffer = (await this.arrayBuffer()) as Buffer;
+
+		return buffer.toString("utf8");
+	}
+
+	stream(): ReadableStream {
+		return Readable.toWeb(createReadStream(this.path));
```
Comment (coyotte508) [inline packages/hub/src/lib/LazyBlob.ts]:
```suggestion
		return Readable.toWeb(createReadStream(this.path, {start: this.start, end: this.end}));
```

**Comment by coyotte508** [comment]
For the TS Error: maybe `as ReturnType<Blob["stream"]>`

**Review by coyotte508 — COMMENTED** [review]
I think it's ready after the changes.

I just can't believe this module doesn't already exist on npm (only versions I found loaded everything in memory)

File: packages/hub/src/lib/LazyBlob.ts [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+		const lazyBlob = new LazyBlob(path, 0, size);
+
+		return lazyBlob;
+	}
+
+	private path: string;
+	private start: number;
+	private end: number;
+	private totalSize: number;
```
Comment (coyotte508) [inline packages/hub/src/lib/LazyBlob.ts:18]: This can be removed as well

File: packages/hub/src/lib/LazyBlob.ts [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+	get size(): number {
+		if (this.start !== null) {
+			if (this.end !== null) {
+				return Math.abs(this.end - this.start);
+			}
+
+			return this.totalSize - this.start;
+		}
+
+		return this.totalSize;
```
Comment (coyotte508) [inline packages/hub/src/lib/LazyBlob.ts:39]:
```suggestion
		return this.end - this.start;
```

File: packages/hub/src/lib/LazyBlob.ts (hunk 10 above)
Comment (Aschen) [reply packages/hub/src/lib/LazyBlob.ts]: I thought about this one on my way back from the beach today :grin:

**Review by coyotte508 — APPROVED** [review]
Alirght, thanks a lot!

**Comment by coyotte508** [comment]
> @coyotte508 What about the documentation? If I understood well it's in the same repo, I can give it a try

I intend to keep it internal. 

 A file with `require("node:fs")` is directly exposed in the exports, it will cause trouble for the use in browsers, at least for the CJS bundle. See how `sha256-node` is currently dynamically imported and only outside the browser.

So I think maybe a smart way to expose it is to allow passing a `URL` as an alternative to a `Blob` - see https://www.geeksforgeeks.org/node-js-url-pathtofileurl-api/ urls and files paths can be equivalent.

It also opens the door for supporting web urls later on as well. (using Range header, or HEAD calls appropriately to get a slice / the size)

**Comment by Aschen** [comment]
Yes I saw, I'm familiar with this kind of dual behavior when it's needed to provide a SDK for both node and the browser (See [Kuzzle Javascript SDK](https://github.com/kuzzleio/sdk-javascript/))

If I understand well, you might want to allows a string representing a path to a local file for `ContentSource` if it's used in Node.js ?

```
await commit({
  repo,
  title: "Some commit",
  credentials: {
    accessToken: TEST_ACCESS_TOKEN,
  },
  operations: [
    {
      operation: "addOrUpdate",
      // 1) Only a path to a local file, if it's used from Node.js, a LazyBlob will be used
      path: "test.txt",
    },
    {
      operation: "addOrUpdate",
      // 2) URL to a local file, if it's used from Node.js, a LazyBlob will be used
      path: "file://test.txt",
    },
  ],
});
```

**Comment by coyotte508** [comment]
> If I understand well, you might want to allows a string representing a path to a local file for `ContentSource` if it's used in Node.js ?

When I was mentioning `URL`, it was literally :)

Basically, change:

```ts
type ContentSource = Blob;
```

To

```ts
type ContentSource = Blob | URL;
```

**Comment by Aschen** [comment]
Ok understood :+1: 

So if the URL is a local file (starting with `file://`) then the `LazyBlob` will be used, am I right?

**Comment by Aschen** [comment]
Do you want me to do this feature, supporting URL as content, in a separate PR?

**Comment by coyotte508** [comment]
> Do you want me to do this feature, supporting URL as content, in a separate PR?

Sure! Then maybe you can remove the export in this one and keep the `LazyBlob` private ? (I was going to do it later today before merging :grin: )
